### PR TITLE
CMake: honour CMAKE_INSTALL_RPATH for Python bindings (fixes #6371)

### LINKED ISSUE #6371: Python bindings do not honor CMAKE_INSTALL_RPATH for CMake build
## Expected behavior and actual behavior.

I build GDAL on Mac OSX and install the bindings to a non-standard location.. By default the Python bindings are installed to `${CMAKE_INSTALL_PREFIX}/lib/python-3.x/site-packages`.  I use `GDAL_PYTHON_INSTALL_PREFIX` to install the Python bindings to a different location. However, the RPATH entry is not updated properly, so `libgdal.31.dylib` is not found without overriding `DYLD_LIBRARY_PATH` at runtime.

I think this could be resolved by updating the Python binding installation to honor [CMAKE_INSTALL_RPATH](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling#always-full-rpath).

## Steps to reproduce the problem.

```bash
cmake .. \
    -DCMAKE_INSTALL_PREFIX=/Users/pschmitt/opt/gdal-3.5.2 \
    -DCMAKE_INSTALL_RPATH=/Users/pschmitt/opt/gdal-3.5.2/lib \
    -DGDAL_PYTHON_INSTALL_PREFIX=$(pyenv prefix)
cmake --build . -j$(sysctl -n hw.ncpu)
cmake --build . --target install
```

`pyenv prefix` evaluates to `/Users/pschmitt/.pyenv/versions/3.8.12`

After install, gdal binaries have the proper rpath set:

```bash
otool -l $(which gdal_translate) | grep -i path
         name @rpath/libgdal.31.dylib (offset 24)
          cmd LC_RPATH
         path /Users/pschmitt/opt/gdal-3.5.2/lib (offset 12)
```

However, the Python bindings do not:

```bash
otool -l /Users/pschmitt/.pyenv/versions/3.8.12/lib/python3.8/site-packages/osgeo/_gdal.cpython-38-darwin.so | grep path
         name @rpath/libgdal.31.dylib (offset 24)
```

This results in runtime errors:

```
python -c "from osgeo import gdal"
Traceback (most recent call last):
  File "/Users/pschmitt/.pyenv/versions/3.8.12/lib/python3.8/site-packages/osgeo/__init__.py", line 30, in swig_import_helper
    return importlib.import_module(mname)
...
ImportError: dlopen(/Users/pschmitt/.pyenv/versions/3.8.12/lib/python3.8/site-packages/osgeo/_gdal.cpython-38-darwin.so, 0x0002): Library not loaded: '@rpath/libgdal.31.dylib'
  Referenced from: '/Users/pschmitt/.pyenv/versions/3.8.12/lib/python3.8/site-packages/osgeo/_gdal.cpython-38-darwin.so'
  Reason: tried: '/usr/local/lib/libgdal.31.dylib' (no such file), '/usr/lib/libgdal.31.dylib' (no such file)
```

## Operating system

macOS Monterrey 12.5.x
python 3.8.12

## GDAL version and provenance

gdal-3.5.2 source

### PR DESCRIPTION
**Pull request by rouault**
but only if it is an absolute path. GDAL_SET_INSTALL_RELATIVE_RPATH=ON will still have no effect on Python bindings.

### CODE REVIEW

File: swig/python/CMakeLists.txt [hunk 1]
```diff
@@ -291,14 +293,57 @@ if (Python_Interpreter_FOUND)
...
     set(BUILD_EXT_FILENAME ${CMAKE_CURRENT_BINARY_DIR}/build_ext.cmake)
+
+    if(${GDAL_SET_INSTALL_RELATIVE_RPATH})
+        # TODO: we should support that, but that's involved as it requires
+        # knowing where Python will install the generated .so files
+        message(WARNING "The Python bindings target does not currently "
+                        "support setting relative RPATH on the Python wrapper "
+                        "shared libraries. "
+                        "Disable GDAL_SET_INSTALL_RELATIVE_RPATH and set "
+                        "an absolute value for CMAKE_INSTALL_PREFIX")
```
Comment (pedros007) [inline swig/python/CMakeLists.txt:304]: Should this be:
```suggestion
                        "an absolute value for CMAKE_INSTALL_RPATH")
```

**Review by pedros007 — COMMENTED** [review]
I get the following error on both `fix_6371` and the `master` branch:

```
cmake --build .
[1/16] Linking CXX executable autotest/pytest_runner
FAILED: autotest/pytest_runner
: && /Library/Developer/CommandLineTools/usr/bin/c++ -fvisibility=hidden -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk -mmacosx-version-min=12.0 -Wl,-search_paths_first -Wl,-headerpad_max_install_names  autotest/CMakeFiles/pytest_runner.dir/pytest_runner.cpp.o -o autotest/pytest_runner  /Users/pschmitt/.pyenv/versions/3.8.12/lib/libpython3.8.a && :
Undefined symbols for architecture x86_64:
  "_libintl_bind_textdomain_codeset", referenced from:
      _PyIntl_bind_textdomain_codeset in libpython3.8.a(_localemodule.o)
  "_libintl_bindtextdomain", referenced from:
      _PyIntl_bindtextdomain in libpython3.8.a(_localemodule.o)
  "_libintl_dcgettext", referenced from:
      _PyIntl_dcgettext in libpython3.8.a(_localemodule.o)
  "_libintl_dgettext", referenced from:
      _PyIntl_dgettext in libpython3.8.a(_localemodule.o)
  "_libintl_gettext", referenced from:
      _PyIntl_gettext in libpython3.8.a(_localemodule.o)
  "_libintl_setlocale", referenced from:
      _PyLocale_setlocale in libpython3.8.a(_localemodule.o)
      _PyLocale_localeconv in libpython3.8.a(_localemodule.o)
  "_libintl_textdomain", referenced from:
      _PyIntl_textdomain in libpython3.8.a(_localemodule.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[2/16] cd /Users/pschmitt/opt/src/gdal-fix_6371/build && /usr/local/Cellar/cmake/3.21.4/bin/cmake -DSOUR...DGDAL_RELEASE_DATE= -P /Users/pschmitt/opt/src/gdal-fix_6371/cmake/helpers/generate_gdal_version_h.cmake
-- Found Git: /usr/bin/git (found version "2.37.0 (Apple Git-136)")
```

using 
```bash
 cmake .. \                                                                                                                                            1 err | 2.7.1 rb
 -GNinja \
 -DCMAKE_OSX_DEPLOYMENT_TARGET=12.0 \
 -DCMAKE_INSTALL_PREFIX=/Users/pschmitt/opt/gdal-fix_6371 \
 -DCMAKE_INSTALL_RPATH=/Users/pschmitt/opt/gdal-fix_6371/lib \
 -DCMAKE_BUILD_TYPE=Release \
 -DEXPAT_INCLUDE_DIR=/usr/local/opt/expat/include \
 -DEXPAT_LIBRARY=/usr/local/opt/expat/lib/libexpat.dylib \
 -DSQLite3_INCLUDE_DIR=/usr/local/opt/sqlite/include \
 -DSQLite3_LIBRARY=/usr/local/opt/sqlite/lib/libsqlite3.dylib \
 -DSPATIALITE_INCLUDE_DIR=/usr/local/opt/libspatialite/include \
 -DSPATIALITE_LIBRARY=/usr/local/opt/libspatialite/lib/libspatialite.dylib \
 -DGDAL_ENABLE_DRIVER_HDF5=OFF \
 -DGDAL_PYTHON_INSTALL_PREFIX=$(pyenv prefix) \
 2>&1 | tee gdal-${GDAL_VERSION}-cmake.log
cmake --build . -j$(sysctl -n hw.ncpu)
```

**Comment by rouault** [comment]
> I get the following error on both fix_6371 and the master branch which I do not see on the 3.5.2 release:

that would be surprising as autotest/CMakeLists.txt is the same in both branches. I'd suspect more something related to your python setup. The error seems to come to the fact that the static library  /Users/pschmitt/.pyenv/versions/3.8.12/lib/libpython3.8.a is used as Python_LIBRARIES

**Comment by pedros007** [comment]
Yes, the changes in this PR seem to solve the `rpath` set in the Python bindings. Thanks!

  The problem I ran into about x86_64 was thanks to homebrew's  gettext package screwing up a pyenv build.  Lately Homebrew seems to cause me more problems than it solves.

**Pull request merged**